src/synapse_api.py

```diff
@@ -0,0 +1,178 @@
+
+    def __init__(self, msg: str):
+        """Initialize a new instance of the NetworkError exception.
+
+        Args:
+            msg (str): Explanation of the error.
+        """
+        self.msg = msg
+
+
+class SynapseAPI:
+    """The Synapse API handler."""
+
+    def __init__(self, charm: CharmBase):
+        """Initialize the handler.
+
+        Args:
+            charm: The parent charm to attach the observer to.
+        """
+        self._charm = charm
+
+    def _get_nonce(self) -> str:
+        """Get nonce.
+
+        Returns:
+            The nonce returned by Synapse API.
+
+        Raises:
+            NetworkError: if there was an error fetching the nonce.
+        """
+        try:
+            res = requests.get(URL_REGISTER, timeout=5)
+            res.raise_for_status()
+            return res.json()["nonce"]
+        except (
+            requests.exceptions.ConnectionError,
+            requests.exceptions.Timeout,
+            requests.exceptions.HTTPError,
+        ) as exc:
+            logger.error("Failed to request %s : %s", URL_REGISTER, exc)
+            raise NetworkError(f"Failed to request {URL_REGISTER}.") from exc
+
+    def _generate_mac(
+        self,
+        shared_secret: str,
+        nonce: str,
+        user: str,
+        password: str,
+        admin: bool = False,
+        user_type: typing.Optional[str] = None,
+    ) -> str:
+        """Generate mac to register user.
+
+        Args:
+            shared_secret: registration_shared_secret from configuration file.
+            nonce: nonce generated.
+            user: username.
+            password: password.
+            admin: if is admin. Default False.
+            user_type: user type. Defaults to None.
+
+        Returns:
+            _type_: _description_
+        """
+        mac = hmac.new(key=shared_secret.encode("utf8"), digestmod=hashlib.sha1)
+        mac.update(nonce.encode("utf8"))
+        mac.update(b"\x00")
+        mac.update(user.encode("utf8"))
+        mac.update(b"\x00")
+        mac.update(password.encode("utf8"))
+        mac.update(b"\x00")
+        mac.update(b"admin" if admin else b"notadmin")
+        if user_type:
+            mac.update(b"\x00")
+            mac.update(user_type.encode("utf8"))
+
+        return mac.hexdigest()
+
+    def register_user(self, username: str, password: str, admin: bool) -> None:
+        """Register user.
+
+        Args:
+            username: name to be registered.
+            password: user's password.
+            admin: if the user is admin or not.
+
+        Raises:
+            RegisterUserError: when registering user via API fails.
+            NetworkError: if there was an error registering the user.
+        """
+        # get registration_shared_secret from config file
+        container = self._charm.unit.get_container(SYNAPSE_CONTAINER_NAME)
+        if not container.can_connect():
+            self._charm.unit.status = ops.MaintenanceStatus("Waiting for pebble")
```

Comment:
This definitely should not be done here, these kind of modules should not be interacting with the charm outside of taking the container as an argument to (1) interact with the file system and (2) execute commands if there isn't an API available

Reply:
Moved to the charm.